File: src/Array/ProductArrayExceptSelf.java
package Array;

// https://leetcode.com/problems/product-of-array-except-self/
// prefix array
public class ProductArrayExceptSelf {
    public int[] productExceptSelf(int[] nums) {
        // Left is an array containing the left products
        // i.e: left[i] = nums[0] * .... * nums[i-1]  * nums[i]
        int[] left = new int[nums.length];

        // Right is an array containing the array products
        //i.e: right[i] = nums[i] * nums[i+1]  * .... * nums[len(nums)]
        int[] right = new int[nums.length];

        // left[i] = left[i-1] * nums[i-1]
        left[0] = 1;
        for (int i = 1; i < nums.length; i++) {
            left[i] = left[i - 1] * nums[i - 1];
        }

        // right[i] = right[i+1] * nums[i+1]
        right[nums.length - 1] = 1;
        for (int i = nums.length - 2; i >= 0; i--) {
            right[i] = right[i + 1] * nums[i + 1];
        }

        int[] product = new int[nums.length];
        for (int i = 0; i < product.length; i++) {
            product[i] = left[i] * right[i];
        }

        return product;
    }

    public static void main(String[] args) {
        ProductArrayExceptSelf p = new ProductArrayExceptSelf();
        int[] input = {1, 2, 3, 4, 5, 6, 7, 8};
        int[] products = p.productExceptSelf(input);
        for (int product :
                products) {
            System.out.print(product + ", ");
        }
    }
}

File: py/essential/ProductArrayExceptSelf.py
# https://leetcode.com/problems/product-of-array-except-self/description/?envType=problem-list-v2&envId=oizxjoit
# Blind
# Grind
# Neet

from typing import List


class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        n = len(nums)
        res = []
        prefixProduct, postfixProducts = [1] * (n + 1), [1] * (
            n + 1
        )  # padding with one more element to avoid index out of range
        for i in range(1, n + 1):
            prefixProduct[i] = prefixProduct[i - 1] * nums[i - 1]
        for i in range(n - 1, -1, -1):
            postfixProducts[i] = postfixProducts[i + 1] * nums[i]
        for i in range(n):
            res.append(prefixProduct[i] * postfixProducts[i + 1])
        return res

    def productExceptSelf2(self, nums: List[int]) -> List[int]:
        # Follow up: Can you solve the problem in O(1) extra space complexity? (The output array does not count as extra space for space complexity analysis.)
        n = len(nums)
        res = [1] * n
        for i in range(1, n):
            res[i] = res[i - 1] * nums[i - 1]
        right = 1
        for i in range(n - 1, -1, -1):
            res[i] *= right
            right *= nums[i]
        return res


s = Solution()
print(s.productExceptSelf2([1, 2, 3, 4, 5]))  # [120,60,40,30,24]
print(s.productExceptSelf2([-1, 1, 0, -3, 3]))  # [0,0,9,0,0]
